src/zeroband/utils/mfu_tracker.py

```
@@ -0,0 +1,332 @@
+
+        assert (N_q == N_k) and (N_k == N_v), "batch size must batch across q, k & v"
+        assert H_k == H_v, "head size must match for keys and values"  # H_q may differ from H_k & H_v
+        assert E_q == E_k, "embedding dim must match for q & k"  # E_v may differ from E_k & E_q
+        assert S_k == S_v, "source sequence length must match for k & v"  # L may differ from S_k & S_v
+
+        N = N_q
+        S = S_k
+        E = E_q
+        mha_flops = 4.0 * N * (S * L) * E
+
+        mask_occupancy = 1.0 - (mask_sparsity.sparsity() / 100.0)
+        forward_flops = math.floor(mha_flops * mask_occupancy)
+        self.track_forward_flops(forward_flops)
+        self.track_backward_flops(math.floor(forward_flops * 2.5))
+
+
+    def track_norm(self, norm: torch.nn.Module, x: torch.Tensor):
+        d = x.size(-1)
+        if isinstance(norm, torch.nn.LayerNorm):
+            if norm.elementwise_affine:
+                if norm.bias is not None:
+                    # With both gamma and beta.
+                    flops = 7 * d + 2
+                else:
+                    # With only gamma scaling.
+                    flops = 6 * d + 2
+            else:
+                flops = 5 * d + 2
+        elif "rmsnorm" in type(norm).__name__.lower():
+            if hasattr(norm, "weight") and norm.weight is not None:
+                flops = 4 * d + 2
+            else:
+                flops = 3 * d + 2
+        else:
+            raise NotImplementedError(f"Normalization type {type(norm)} not supported for flop tracking.")
+
+        self.track_forward_flops(flops)
+
+    def track_optimizer_step(self, optimizer: torch.optim.Optimizer, num_param_scalars: int):
+        if isinstance(optimizer, torch.optim.Adam):
+            flops_per_param = 14
+        elif isinstance(optimizer, torch.optim.AdamW):
+            flops_per_param = 16
+        else:
+            raise NotImplementedError(f"Optimizer type {type(optimizer)} not supported for flop tracking.")
```

lets not break if the optimizer is not this one and rather fall back to adamw count + warning

When working on a new optimizer we would add the flops counter only as a the last step

---

yeah, but the silent flop tracking error is something that you probably want to avoid.
This is at the other end of the code base and you don't necessarily expect that you have to change this part to *actually* make it work.
If you change the optimizer, it should fail once such that that part of the code base is in your mind going forward, you either add the branch right then and there or you hack it, knowing that you eventually have to change it.
But I would suggest we don't silently fail.